File: backend/app/schemas/evaluation.py
from datetime import date, datetime
from typing import Any
from typing import Literal

from pydantic import BaseModel, EmailStr, Field


CycleStatus = Literal["RASCUNHO", "EM_COLETA", "EM_ANALISE", "EM_CALIBRACAO", "FINALIZADO"]
RelationType = Literal["MANAGER", "PEER", "INTERNAL_CLIENT", "SELF"]
EvaluationCategory = Literal["DESTAQUE", "MUITO_BOM", "BOM", "EM_DESENVOLVIMENTO", "ATENCAO"]
AlertType = Literal[
    "HIGH_PERFORMANCE_LOW_BEHAVIOR",
    "LOW_PERFORMANCE_HIGH_BEHAVIOR",
    "MANAGER_PEER_DIVERGENCE",
    "MISSING_DATA",
]
AlertSeverity = Literal["LOW", "MEDIUM", "HIGH"]
ImportStatus = Literal["UPLOADED", "MAPPED", "VALIDATED", "IMPORTED", "ERROR"]
ImportRowStatus = Literal["PENDING", "VALID", "IMPORTED", "ERROR"]
AiAnalysisStatus = Literal["PENDING", "PROCESSED", "REVIEWED", "ERROR"]


class EvaluationCycleBase(BaseModel):
    name: str = Field(min_length=2, max_length=200)
    start_date: date
    end_date: date
    status: CycleStatus = "RASCUNHO"
    performance_weight: float = Field(default=0.10, ge=0, le=1)
    behavior_weight: float = Field(default=0.45, ge=0, le=1)
    potential_weight: float = Field(default=0.45, ge=0, le=1)
    notes: str | None = None


class EvaluationCycleCreate(EvaluationCycleBase):
    pass


class EvaluationCycleUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=200)
    start_date: date | None = None
    end_date: date | None = None
    performance_weight: float | None = Field(default=None, ge=0, le=1)
    behavior_weight: float | None = Field(default=None, ge=0, le=1)
    potential_weight: float | None = Field(default=None, ge=0, le=1)
    notes: str | None = None


class EvaluationCycleStatusUpdate(BaseModel):
    status: CycleStatus


class EvaluationCycleOut(EvaluationCycleBase):
    id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class EmployeeBase(BaseModel):
    name: str = Field(min_length=2, max_length=200)
    email: EmailStr
    teams_user_id: str | None = None
    matricula: str | None = None
    cargo: str | None = None
    setor: str | None = None
    department: str | None = None
    position: str | None = None
    manager_id: int | None = None
    active: bool = True
    recebe_notificacao: bool = True
    participa_avaliacao: bool = True
    canal_preferencial: str = "email"


class EmployeeCreate(EmployeeBase):
    pass


class EmployeeUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=200)
    email: EmailStr | None = None
    teams_user_id: str | None = None
    matricula: str | None = None
    cargo: str | None = None
    setor: str | None = None
    department: str | None = None
    position: str | None = None
    manager_id: int | None = None
    active: bool | None = None
    recebe_notificacao: bool | None = None
    participa_avaliacao: bool | None = None
    canal_preferencial: str | None = None


class EmployeeOut(EmployeeBase):
    id: int
    manager_name: str | None = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class IndicatorPayload(BaseModel):
    rpm_original: float | None = Field(default=None, ge=0)
    ihpe_original: float | None = Field(default=None, ge=0)


class IndicatorOut(BaseModel):
    id: int
    cycle_id: int
    employee_id: int
    employee_name: str | None = None
    rpm_original: float | None
    rpm_normalized: float | None
    ihpe_original: float | None
    ihpe_normalized: float | None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class OperationalImportOut(BaseModel):
    imported_rows: int
    updated_employees: int
    updated_indicators: int = 0
    updated_rh_records: int = 0
    warnings: list[str] = []


class EmployeeRhDataOut(BaseModel):
    id: int
    cycle_id: int
    employee_id: int
    employee_name: str | None = None
    career_level: int | None
    last_merit_date: date | None
    admission_date: date | None
    is_level_one_separate_budget: bool
    eligible_for_merit: bool
    eligibility_reason: str | None

    class Config:
        from_attributes = True


class Review360Create(BaseModel):
    evaluator_id: int | None = None
    evaluated_id: int
    evaluator_email: EmailStr | None = None
    evaluator_name: str | None = None
    evaluated_email: EmailStr | None = None
    evaluated_name: str | None = None
    relation_type: RelationType
    score: float | None = Field(default=None, ge=0, le=100)
    general_score: float | None = Field(default=None, ge=0, le=100)
    communication_score: float | None = Field(default=None, ge=0, le=100)
    teamwork_score: float | None = Field(default=None, ge=0, le=100)
    commitment_score: float | None = Field(default=None, ge=0, le=100)
    autonomy_score: float | None = Field(default=None, ge=0, le=100)
    quality_score: float | None = Field(default=None, ge=0, le=100)
    problem_solving_score: float | None = Field(default=None, ge=0, le=100)
    strengths_comment: str | None = None
    improvement_comment: str | None = None
    general_comment: str | None = None
    comment: str | None = None


class Review360Out(Review360Create):
    id: int
    cycle_id: int
    import_id: int | None = None
    import_row_id: int | None = None
    evaluator_name: str | None = None
    evaluated_name: str | None = None
    submitted_at: datetime
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PotentialPayload(BaseModel):
    score: float = Field(ge=0, le=100)
    comment: str | None = None


class PotentialOut(BaseModel):
    id: int
    cycle_id: int
    employee_id: int
    employee_name: str | None = None
    score: float
    comment: str | None
    created_by: int | None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class EvaluationAlertOut(BaseModel):
    id: int
    cycle_id: int
    employee_id: int
    employee_name: str | None = None
    alert_type: AlertType
    message: str
    severity: AlertSeverity
    created_at: datetime
    resolved_at: datetime | None
    resolved_by: int | None

    class Config:
        from_attributes = True


class EvaluationScoreOut(BaseModel):
    id: int
    cycle_id: int
    employee_id: int
    employee_name: str
    department: str | None
    position: str | None
    manager_name: str | None = None
    performance_score: float | None
    behavior_score: float | None
    potential_score: float | None
    preliminary_final_score: float | None
    suggested_category: EvaluationCategory | None
    final_category: EvaluationCategory | None
    nine_box_position: str | None
    calibration_justification: str | None
    calibrated_by: int | None
    calibrated_at: datetime | None
    alerts: list[EvaluationAlertOut] = []

    class Config:
        from_attributes = True


class CalculateScoresOut(BaseModel):
    cycle_id: int
    processed: int
    incomplete: int
    alerts_generated: int


class CalibrationPayload(BaseModel):
    final_category: EvaluationCategory
    calibration_justification: str | None = None


class FinalListRowOut(EvaluationScoreOut):
    position: int


class EvaluationImportOut(BaseModel):
    id: int
    cycle_id: int
    file_name: str
    status: ImportStatus
    uploaded_by: int | None
    uploaded_at: datetime
    column_mapping_json: dict[str, str] | None
    total_rows: int
    valid_rows: int
    invalid_rows: int
    error_message: str | None
    headers: list[str] = []

    class Config:
        from_attributes = True


class ImportColumnMappingPayload(BaseModel):
    mapping: dict[str, str]


class EvaluationImportRowOut(BaseModel):
    id: int
    import_id: int
    row_number: int
    raw_data_json: dict[str, Any]
    normalized_data_json: dict[str, Any] | None
    status: ImportRowStatus
    error_message: str | None

    class Config:
        from_attributes = True


class ImportValidationOut(BaseModel):
    import_id: int
    status: ImportStatus
    total_rows: int
    valid_rows: int
    invalid_rows: int
    errors: list[EvaluationImportRowOut]


class ImportConfirmOut(BaseModel):
    import_id: int
    imported_rows: int
    created_reviews: int


class QualitativeAlertOut(BaseModel):
    type: str
    description: str
    severity: AlertSeverity


class AiFeedbackAnalysisOut(BaseModel):
    id: int
    cycle_id: int
    employee_id: int
    employee_name: str | None = None
    status: AiAnalysisStatus
    summary: str | None
    strengths_json: list[str] | None
    attention_points_json: list[str] | None
    recurring_themes_json: list[str] | None
    qualitative_alerts_json: list[dict[str, Any]] | None
    suggested_feedback: str | None
    model_used: str | None
    raw_response_json: dict[str, Any] | None
    error_message: str | None
    reviewed_by: int | None
    reviewed_at: datetime | None

    class Config:
        from_attributes = True


class AiAnalysisReviewPayload(BaseModel):
    reviewed: bool = True


class PreliminaryReportOut(BaseModel):
    employee: EmployeeOut
    score: EvaluationScoreOut | None
    reviews: list[Review360Out]
    ai_analysis: AiFeedbackAnalysisOut | None
    alerts: list[EvaluationAlertOut]


class DashboardOut(BaseModel):
    cycle_id: int
    total_evaluated: int
    average_final_score: float | None
    suggested_by_category: dict[str, int]
    final_by_category: dict[str, int]
    by_department: dict[str, int]
    alerts: list[EvaluationAlertOut]

File: frontend/src/api/evaluation.ts
import api from './client';

export type CycleStatus = 'RASCUNHO' | 'EM_COLETA' | 'EM_ANALISE' | 'EM_CALIBRACAO' | 'FINALIZADO';
export type EvaluationCategory = 'DESTAQUE' | 'MUITO_BOM' | 'BOM' | 'EM_DESENVOLVIMENTO' | 'ATENCAO';
export type RelationType = 'MANAGER' | 'PEER' | 'INTERNAL_CLIENT' | 'SELF';
export type ImportStatus = 'UPLOADED' | 'MAPPED' | 'VALIDATED' | 'IMPORTED' | 'ERROR';

export interface EvaluationCycle {
  id: number;
  name: string;
  start_date: string;
  end_date: string;
  status: CycleStatus;
  performance_weight: number;
  behavior_weight: number;
  potential_weight: number;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface Employee {
  id: number;
  name: string;
  email: string;
  teams_user_id?: string | null;
  matricula?: string | null;
  cargo?: string | null;
  setor?: string | null;
  department: string | null;
  position: string | null;
  manager_id: number | null;
  manager_name: string | null;
  active: boolean;
  recebe_notificacao?: boolean;
  participa_avaliacao?: boolean;
  canal_preferencial?: string;
  created_at: string;
  updated_at: string;
}

export interface Indicator {
  id: number;
  cycle_id: number;
  employee_id: number;
  employee_name: string | null;
  rpm_original: number | null;
  rpm_normalized: number | null;
  ihpe_original: number | null;
  ihpe_normalized: number | null;
}

export interface EvaluationAlert {
  id: number;
  cycle_id: number;
  employee_id: number;
  employee_name: string | null;
  alert_type: string;
  message: string;
  severity: 'LOW' | 'MEDIUM' | 'HIGH';
  created_at: string;
  resolved_at: string | null;
  resolved_by: number | null;
}

export interface EvaluationScoreRow {
  id: number;
  cycle_id: number;
  employee_id: number;
  employee_name: string;
  department: string | null;
  position: string | null;
  manager_name: string | null;
  performance_score: number | null;
  behavior_score: number | null;
  potential_score: number | null;
  preliminary_final_score: number | null;
  suggested_category: EvaluationCategory | null;
  final_category: EvaluationCategory | null;
  nine_box_position: string | null;
  calibration_justification: string | null;
  calibrated_by: number | null;
  calibrated_at: string | null;
  alerts: EvaluationAlert[];
}

export interface Review360 {
  id: number;
  cycle_id: number;
  import_id?: number | null;
  import_row_id?: number | null;
  evaluator_id: number | null;
  evaluator_email?: string | null;
  evaluator_name: string | null;
  evaluated_id: number;
  evaluated_email?: string | null;
  evaluated_name: string | null;
  relation_type: RelationType;
  score: number | null;
  general_score?: number | null;
  communication_score?: number | null;
  teamwork_score?: number | null;
  commitment_score?: number | null;
  autonomy_score?: number | null;
  quality_score?: number | null;
  problem_solving_score?: number | null;
  strengths_comment?: string | null;
  improvement_comment?: string | null;
  general_comment?: string | null;
  comment: string | null;
}

export interface Potential {
  id: number;
  cycle_id: number;
  employee_id: number;
  employee_name: string | null;
  score: number;
  comment: string | null;
}

export interface CalculateResult {
  cycle_id: number;
  processed: number;
  incomplete: number;
  alerts_generated: number;
}

export interface Dashboard {
  cycle_id: number;
  total_evaluated: number;
  average_final_score: number | null;
  suggested_by_category: Record<string, number>;
  final_by_category: Record<string, number>;
  by_department: Record<string, number>;
  alerts: EvaluationAlert[];
}

export interface EvaluationImport {
  id: number;
  cycle_id: number;
  file_name: string;
  status: ImportStatus;
  uploaded_by: number | null;
  uploaded_at: string;
  column_mapping_json: Record<string, string> | null;
  total_rows: number;
  valid_rows: number;
  invalid_rows: number;
  error_message: string | null;
  headers: string[];
}

export interface EvaluationImportRow {
  id: number;
  import_id: number;
  row_number: number;
  raw_data_json: Record<string, unknown>;
  normalized_data_json: Record<string, unknown> | null;
  status: 'PENDING' | 'VALID' | 'IMPORTED' | 'ERROR';
  error_message: string | null;
}

export interface ImportValidationResult {
  import_id: number;
  status: ImportStatus;
  total_rows: number;
  valid_rows: number;
  invalid_rows: number;
  errors: EvaluationImportRow[];
}

export interface ImportConfirmResult {
  import_id: number;
  imported_rows: number;
  created_reviews: number;
}

export interface OperationalImportResult {
  imported_rows: number;
  updated_employees: number;
  updated_indicators: number;
  updated_rh_records: number;
  warnings: string[];
}

export interface AiFeedbackAnalysis {
  id: number;
  cycle_id: number;
  employee_id: number;
  employee_name: string | null;
  status: 'PENDING' | 'PROCESSED' | 'REVIEWED' | 'ERROR';
  summary: string | null;
  strengths_json: string[] | null;
  attention_points_json: string[] | null;
  recurring_themes_json: string[] | null;
  qualitative_alerts_json: Array<Record<string, unknown>> | null;
  suggested_feedback: string | null;
  model_used: string | null;
  error_message: string | null;
}

export interface PreliminaryReport {
  employee: Employee;
  score: EvaluationScoreRow | null;
  reviews: Review360[];
  ai_analysis: AiFeedbackAnalysis | null;
  alerts: EvaluationAlert[];
}

export const listEvaluationCycles = async () => (await api.get<EvaluationCycle[]>('/evaluation-cycles')).data;

export const createEvaluationCycle = async (payload: {
  name: string;
  start_date: string;
  end_date: string;
  status?: CycleStatus;
  notes?: string | null;
}) => (await api.post<EvaluationCycle>('/evaluation-cycles', payload)).data;

export const updateEvaluationCycleStatus = async (id: number, status: CycleStatus) =>
  (await api.patch<EvaluationCycle>(`/evaluation-cycles/${id}/status`, { status })).data;

export const listEmployees = async (params?: { department?: string; manager_id?: number }) =>
  (await api.get<Employee[]>('/employees', { params })).data;

export const createEmployee = async (payload: {
  name: string;
  email: string;
  teams_user_id?: string | null;
  matricula?: string | null;
  cargo?: string | null;
  setor?: string | null;
  department?: string | null;
  position?: string | null;
  manager_id?: number | null;
  active?: boolean;
  recebe_notificacao?: boolean;
  participa_avaliacao?: boolean;
  canal_preferencial?: string;
}) => (await api.post<Employee>('/employees', payload)).data;

export const updateEmployee = async (id: number, payload: Partial<Employee>) =>
  (await api.put<Employee>(`/employees/${id}`, payload)).data;

export const uploadEvaluationCsv = async (cycleId: number, file: File) => {
  const formData = new FormData();
  formData.append('file', file);
  return (await api.post<EvaluationImport>(`/evaluation-cycles/${cycleId}/imports`, formData)).data;
};

export const listEvaluationImports = async (cycleId: number) =>
  (await api.get<EvaluationImport[]>(`/evaluation-cycles/${cycleId}/imports`)).data;

export const mapImportColumns = async (cycleId: number, importId: number, mapping: Record<string, string>) =>
  (await api.post<EvaluationImport>(`/evaluation-cycles/${cycleId}/imports/${importId}/map-columns`, { mapping })).data;

export const validateImport = async (cycleId: number, importId: number) =>
  (await api.post<ImportValidationResult>(`/evaluation-cycles/${cycleId}/imports/${importId}/validate`)).data;

export const confirmImport = async (cycleId: number, importId: number) =>
  (await api.post<ImportConfirmResult>(`/evaluation-cycles/${cycleId}/imports/${importId}/confirm`)).data;

export const getImportErrors = async (cycleId: number, importId: number) =>
  (await api.get<EvaluationImportRow[]>(`/evaluation-cycles/${cycleId}/imports/${importId}/errors`)).data;

export const listIndicators = async (cycleId: number) =>
  (await api.get<Indicator[]>(`/evaluation-cycles/${cycleId}/indicators`)).data;

export const upsertIndicator = async (cycleId: number, employeeId: number, payload: { rpm_original: number | null; ihpe_original: number | null }) =>
  (await api.put<Indicator>(`/evaluation-cycles/${cycleId}/employees/${employeeId}/indicators`, payload)).data;

export const listReviews = async (cycleId: number) =>
  (await api.get<Review360[]>(`/evaluation-cycles/${cycleId}/reviews`)).data;

export const createReview = async (cycleId: number, payload: { evaluator_id: number; evaluated_id: number; relation_type: RelationType; score: number; comment?: string | null }) =>
  (await api.post<Review360>(`/evaluation-cycles/${cycleId}/reviews`, payload)).data;

export const upsertPotential = async (cycleId: number, employeeId: number, payload: { score: number; comment?: string | null }) =>
  (await api.put<Potential>(`/evaluation-cycles/${cycleId}/employees/${employeeId}/potential`, payload)).data;

export const calculateScores = async (cycleId: number) =>
  (await api.post<CalculateResult>(`/evaluation-cycles/${cycleId}/calculate-scores`)).data;

export const listScores = async (cycleId: number) =>
  (await api.get<EvaluationScoreRow[]>(`/evaluation-cycles/${cycleId}/scores`)).data;

export const runCycleAiAnalysis = async (cycleId: number) =>
  (await api.post<AiFeedbackAnalysis[]>(`/evaluation-cycles/${cycleId}/run-ai-analysis`)).data;

export const runEmployeeAiAnalysis = async (cycleId: number, employeeId: number) =>
  (await api.post<AiFeedbackAnalysis>(`/evaluation-cycles/${cycleId}/employees/${employeeId}/run-ai-analysis`)).data;

export const getEmployeeAiAnalysis = async (cycleId: number, employeeId: number) =>
  (await api.get<AiFeedbackAnalysis>(`/evaluation-cycles/${cycleId}/employees/${employeeId}/ai-analysis`)).data;

export const getPreliminaryReport = async (cycleId: number, employeeId: number) =>
  (await api.get<PreliminaryReport>(`/evaluation-cycles/${cycleId}/employees/${employeeId}/preliminary-report`)).data;

export const getCalibration = async (cycleId: number) =>
  (await api.get<EvaluationScoreRow[]>(`/evaluation-cycles/${cycleId}/calibration`)).data;

export const calibrateEmployee = async (cycleId: number, employeeId: number, payload: { final_category: EvaluationCategory; calibration_justification?: string | null }) =>
  (await api.patch<EvaluationScoreRow>(`/evaluation-cycles/${cycleId}/employees/${employeeId}/calibration`, payload)).data;

export const getFinalList = async (cycleId: number, params?: { department?: string; category?: string }) =>
  (await api.get<EvaluationScoreRow[]>(`/evaluation-cycles/${cycleId}/final-list`, { params })).data;

export const exportFinalListCsv = async (cycleId: number) =>
  (await api.get(`/evaluation-cycles/${cycleId}/final-list/export`, { responseType: 'blob' })).data as Blob;

export const getFinalReportHtml = async (cycleId: number) =>
  (await api.get(`/evaluation-cycles/${cycleId}/final-report/html`, { responseType: 'blob' })).data as Blob;

const uploadOperationalFile = async (cycleId: number, kind: 'rpm' | 'ihpe' | 'rh', file: File) => {
  const formData = new FormData();
  formData.append('file', file);
  return (await api.post<OperationalImportResult>(`/evaluation-cycles/${cycleId}/operational-imports/${kind}`, formData)).data;
};

export const uploadRpmFile = async (cycleId: number, file: File) => uploadOperationalFile(cycleId, 'rpm', file);
export const uploadIhpeFile = async (cycleId: number, file: File) => uploadOperationalFile(cycleId, 'ihpe', file);
export const uploadRhFile = async (cycleId: number, file: File) => uploadOperationalFile(cycleId, 'rh', file);

export const listAlerts = async (cycleId: number) =>
  (await api.get<EvaluationAlert[]>(`/evaluation-cycles/${cycleId}/alerts`)).data;

export const getDashboard = async (cycleId: number) =>
  (await api.get<Dashboard>(`/evaluation-cycles/${cycleId}/dashboard`)).data;
